### PR TITLE
OSA external API

### PR DESCRIPTION
**Pull request by pweil-**
Continuation of https://github.com/Azure/acs-engine/pull/3434 while @jim-minter is out

@pweil- @kargakis @mjudeikis @amanohar @qike-ms @nicolerenee ptal

I will hold on vendoring anything back in to azure-helm until we're more certain about this shape.

Initial feedback includes: separation of compute and infra pools, pointers, updated validation

### CODE REVIEW

File: pkg/api/osa/vlabs/convert.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// api.ContainerService.
+func (oc *OpenShiftCluster) AsContainerService() *api.ContainerService {
+	cs := &api.ContainerService{
+		ID:       oc.ID,
+		Location: oc.Location,
+		Name:     oc.Name,
+		Tags:     oc.Tags,
+		Type:     oc.Type,
+		Properties: &api.Properties{
+			ProvisioningState: api.ProvisioningState(oc.Properties.ProvisioningState),
```
Comment (kargakis) [inline pkg/api/osa/vlabs/convert.go]: Remind me why we need this? A comment would be nice.

File: pkg/api/osa/vlabs/types.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// AgentPoolProfile represents configuration of VMs running agent daemons that
+// register with the master and offer resources to host applications in
+// containers.
+type AgentPoolProfile struct {
+	Name         string `json:"name,omitempty"`
+	Count        int    `json:"count,omitempty"`
+	VMSize       string `json:"vmSize,omitempty"`
+	VnetSubnetID string `json:"vnetSubnetID,omitempty"`
+	// OSDiskSizeGB int `json:"osDiskSizeGB,omitempty"` // TODO: do we need to add this?
+	// AvailabilityProfile string `json:"availabilityProfile,omitempty"` // TODO: do we need to add this?
```
Comment (kargakis) [inline pkg/api/osa/vlabs/types.go]: I expect no

File: pkg/api/osa/vlabs/types.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	// StorageProfile string `json:"storageProfile,omitempty"` // TODO: do we need to add this?
```
Comment (kargakis) [inline pkg/api/osa/vlabs/types.go]: No, only managed disks

File: pkg/api/osa/vlabs/types.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	// OSType OSType `json:"osType,omitempty"` // TODO: do we need to add this?
```
Comment (kargakis) [inline pkg/api/osa/vlabs/types.go]: Hrm, eventually it seems we will need this (unclear whether it's a blocker for GA).

File: pkg/api/osa/vlabs/types.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// AgentPoolProfile represents configuration of VMs running agent daemons that
```
Comment (kargakis) [inline pkg/api/osa/vlabs/types.go]: s/AgentPoolProfile/InfraPoolProfile/

File: pkg/api/osa/vlabs/validate.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	default:
+		return fmt.Errorf("invalid provisioningState %q", p.ProvisioningState)
+	}
+
+	if p.OpenShiftVersion == "" {
+		return fmt.Errorf("openShiftVersion must not be empty")
+	}
+
+	if p.PublicHostname == "" {
+		return fmt.Errorf("publicHostname must not be empty")
```
Comment (kargakis) [inline pkg/api/osa/vlabs/validate.go]: Move in `isValidHostname`

File: pkg/api/osa/vlabs/validate.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		if err := app.Validate(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := infra.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
```
Comment (kargakis) [inline pkg/api/osa/vlabs/validate.go]: return errors.NewAggregate(errs)

File: pkg/api/osa/vlabs/validate.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		return errors.NewAggregate(errs)
+	}
+
+	return nil
+}
+
+// Validate validates an InfraPoolProfile.
+func (infra *InfraPoolProfile) Validate() error {
+	errs := []error{}
```
Comment (kargakis) [inline pkg/api/osa/vlabs/validate.go]: var errs []error

File: pkg/api/osa/vlabs/validate.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+	if !regexAgentPoolName.MatchString(infra.Name) {
+		return fmt.Errorf("invalid name %q", infra.Name)
+	}
+
+	if infra.Count < 3 {
+		errs = append(errs, fmt.Errorf("must have at least 3 infra nodes"))
```
Comment (kargakis) [inline pkg/api/osa/vlabs/validate.go]: Unclear why we would force HA infra profile.

File: pkg/api/osa/vlabs/types.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+// register with the master and offer resources to host infra applications in
+// containers.
+//
+// NOTE: right now this is just a copy of AgentPoolProfile.  It allows us to
+// control the settings in a way specific to infra nodes if we need to in the future.
+type InfraPoolProfile struct {
+	Name         string `json:"name,omitempty"`
```
Comment (kargakis) [inline pkg/api/osa/vlabs/types.go]: Embed AgentPoolProfile

**Comment by kargakis** [comment]
Can you forsee any additional knobs we will want to enable for infra nodes?

File: pkg/api/osa/vlabs/validate.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+	return p.ServicePrincipalProfile.Validate()
+}
+
+func ValidatePools(compute AgentPoolProfiles, infra *InfraPoolProfile) error {
+	if infra == nil || len(compute) == 0 {
+		return fmt.Errorf("both infra and compute pools are required")
+	}
+
+	errs := []error{}
```
Comment (kargakis) [inline pkg/api/osa/vlabs/validate.go]: var errs []error

File: pkg/api/osa/vlabs/types.go (hunk 4 above)
Comment (pweil-) [reply pkg/api/osa/vlabs/types.go]: non-blocker.

File: pkg/api/osa/vlabs/validate.go (hunk 9 above)
Comment (pweil-) [reply pkg/api/osa/vlabs/validate.go]: we should not leave it up to the user

File: pkg/api/osa/vlabs/types.go (hunk 10 above)
Comment (pweil-) [reply pkg/api/osa/vlabs/types.go]: That is what we're trying *not* to do by separating the profiles (comments on https://github.com/Azure/acs-engine/pull/3434).  The idea/suggestion was that it is possible we want far less options in the infra pool profile than the agent pool profile.  cc @amanohar

File: pkg/api/osa/vlabs/validate.go (hunk 9 above)
Comment (kargakis) [reply pkg/api/osa/vlabs/validate.go]: Agreed. I was thinking of profiles: non-HA (1), HA (3+)

File: pkg/api/osa/vlabs/types.go (hunk 10 above)
Comment (kargakis) [reply pkg/api/osa/vlabs/types.go]: sgtm

**Comment by pweil-** [comment]
> Can you forsee any additional knobs we will want to enable for infra nodes?

Not more, but maybe less.  For instance, do we allow the user to choose the name?  Not much point in doing so.

**Comment by pweil-** [comment]
@kargakis feedback in round2 commit

File: pkg/api/osa/vlabs/convert.go (hunk 1 above)
Comment (pweil-) [reply pkg/api/osa/vlabs/convert.go]: I would rather see this as a status struct (that also has condition fields) but this is the way other apis are reporting status.  I assume that there is reporting somewhere based on it.  Maybe @amanohar @qike-ms @nicolerenee  can tell us more

Comment (amanohar) [reply pkg/api/osa/vlabs/convert.go]: @kargakis @pweil- here is more context about ARM's `provisioningState` field: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property

Detailed error is returned by RP to ARM via operation resource: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property

File: pkg/api/osa/vlabs/types.go (hunk 4 above)
Comment (jim-minter) [reply pkg/api/osa/vlabs/types.go]: intend to add this in as it's likely enough + neater to add from the beginning than retrofit

File: pkg/api/osa/vlabs/validate.go (hunk 6 above)
Comment (jim-minter) [reply pkg/api/osa/vlabs/validate.go]: not doing this because of the change in semantic of the field

File: pkg/api/osa/vlabs/validate.go (hunk 9 above)
Comment (jim-minter) [reply pkg/api/osa/vlabs/validate.go]: I've been putting this kind of thing in plugin validation, not generic validation

File: pkg/api/osa/vlabs/types.go (hunk 10 above)
Comment (jim-minter) [reply pkg/api/osa/vlabs/types.go]: I'd like to roll this back - if we're going to have compute, infra and master and who knows what set of user-allowed configurables I think they're best in one AgentPoolProfile struct not multiple

File: pkg/api/osa/vlabs/validate.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
+package vlabs
+
+import (
+	"fmt"
+	"regexp"
+
+	"k8s.io/apimachinery/pkg/util/errors"
```
Comment (jim-minter) [inline pkg/api/osa/vlabs/validate.go]: IMO doing this is very bad style in an api package.

**Comment by jim-minter** [comment]
Closing for https://github.com/Azure/acs-engine/pull/3556

**Pull request closed without merging**